File: src/packages/control-ui/src/api/split_tunnel_rule_sync.ts
// Control UI TypeScript: Split Tunnel Routing Rule Synchronizer API

export type SplitRoutingAction = 'route_through_vpn' | 'bypass_vpn' | 'drop_traffic';

export interface SplitRule {
  id: string;
  networkStr: string;
  prefixLen: number;
  action: SplitRoutingAction;
  priority: number;
}

export class SplitTunnelRuleSync {
  public defaultAction: SplitRoutingAction;
  public rules: SplitRule[] = [];
  public syncVersion: number = 0;

  constructor(defaultAction: SplitRoutingAction = 'bypass_vpn') {
    this.defaultAction = defaultAction;
  }

  public addRule(
    id: string,
    cidr: string,
    action: SplitRoutingAction,
    priority: number = 100,
  ): void {
    const parts = cidr.trim().split('/');
    const netStr = parts[0]!;
    const prefix = parts.length === 2 ? parseInt(parts[1]!, 10) : 32;

    this.rules.push({
      id,
      networkStr: netStr,
      prefixLen: prefix,
      action,
      priority,
    });

    this.rules.sort((a, b) => {
      if (b.priority !== a.priority) return b.priority - a.priority;
      return b.prefixLen - a.prefixLen;
    });

    this.syncVersion++;
  }

  public matchIp(ip: string): SplitRoutingAction {
    for (const rule of this.rules) {
      if (this.isIpInSubnet(ip, rule.networkStr, rule.prefixLen)) {
        return rule.action;
      }
    }
    return this.defaultAction;
  }

  public syncFromFeed(feed: string, action: SplitRoutingAction): number {
    let count = 0;
    for (const line of feed.split('\n')) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith('#')) continue;
      this.addRule(`feed-${count + 1}`, trimmed, action, 100);
      count++;
    }
    return count;
  }

  private isIpInSubnet(ip: string, network: string, prefix: number): boolean {
    const ipNum = this.ipToNum(ip);
    const netNum = this.ipToNum(network);
    const mask = prefix === 0 ? 0 : (~0 << (32 - prefix)) >>> 0;
    return (ipNum & mask) === (netNum & mask);
  }

  private ipToNum(ip: string): number {
    return ip.split('.').reduce((acc, octet) => ((acc << 8) + parseInt(octet, 10)) >>> 0, 0);
  }
}

File: labs/mobile/donor-archive/tunnel/SplitTunnelRuleSync.kt
package com.luminet.android.tunnel

import java.net.InetAddress
import java.util.concurrent.CopyOnWriteArrayList

enum class AndroidSplitAction {
    ROUTE_THROUGH_VPN,
    BYPASS_VPN,
    DROP_TRAFFIC
}

data class AndroidSplitRule(
    val id: String,
    val networkStr: String,
    val prefixLen: Int,
    val action: AndroidSplitAction,
    val priority: Int
)

class SplitTunnelRuleSync(val defaultAction: AndroidSplitAction = AndroidSplitAction.BYPASS_VPN) {
    val rules = CopyOnWriteArrayList<AndroidSplitRule>()
    var syncVersion: Long = 0
        private set

    fun addRule(id: String, cidr: String, action: AndroidSplitAction, priority: Int) {
        val parts = cidr.trim().split("/")
        val netStr = parts[0]
        val prefix = if (parts.size == 2) parts[1].toInt() else 32

        val rule = AndroidSplitRule(id, netStr, prefix, action, priority)
        rules.add(rule)
        rules.sortWith(compareByDescending<AndroidSplitRule> { it.priority }.thenByDescending { it.prefixLen })
        syncVersion++
    }

    fun matchIp(destIp: String): AndroidSplitAction {
        for (rule in rules) {
            if (matchesCidr(destIp, rule.networkStr, rule.prefixLen)) {
                return rule.action
            }
        }
        return defaultAction
    }

    fun syncFromFeed(feedContent: String, action: AndroidSplitAction): Int {
        var count = 0
        for (line in feedContent.lines()) {
            val trimmed = line.trim()
            if (trimmed.isEmpty() || trimmed.startsWith("#")) continue
            addRule("feed-${count + 1}", trimmed, action, 100)
            count++
        }
        return count
    }

    private fun matchesCidr(ipStr: String, netStr: String, prefix: Int): Boolean {
        return try {
            val netAddr = InetAddress.getByName(netStr).address
            val targetAddr = InetAddress.getByName(ipStr).address
            if (prefix == 0) return true

            var matched = true
            var bits = prefix
            for (i in netAddr.indices) {
                if (bits >= 8) {
                    if (netAddr[i] != targetAddr[i]) { matched = false; break }
                    bits -= 8
                } else if (bits > 0) {
                    val mask = (0xFF shl (8 - bits)) and 0xFF
                    if ((netAddr[i].toInt() and mask) != (targetAddr[i].toInt() and mask)) { matched = false; break }
                    bits = 0
                } else break
            }
            matched
        } catch (e: Exception) {
            false
        }
    }
}
